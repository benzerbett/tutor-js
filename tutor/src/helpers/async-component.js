import { asyncComponent as asyncLoader } from 'react-async-component';
import ErrorComponent from '../components/error-monitoring/async-load-error';
import LoadingComponent from '../components/loading-screen';

export function asyncComponent(resolve) {
  return asyncLoader({
    resolve,
    LoadingComponent,
    ErrorComponent,
  });
}

export function loadAsync(resolve) {
  // return a function so the router will only evaluate it when it's needed
  // in the future we can insert role dependant logic here
  return () => asyncComponent(resolve);
}
