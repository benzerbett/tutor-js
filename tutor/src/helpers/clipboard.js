import _ from 'underscore';

const SUPPORTED = typeof document.queryCommandSupported === 'function' ? document.queryCommandSupported('copy') : undefined;

export default {
  isSupported: _.constant(!!SUPPORTED),

  copy() {
    if (!SUPPORTED) { return; }

    try {
      return document.execCommand('copy');
    } catch (e) {
      return console.warn('clipboard copy failed', e);
    }
  },
};
