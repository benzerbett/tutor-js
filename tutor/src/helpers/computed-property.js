import { forIn } from 'lodash';
import { computed } from 'mobx';

export function addComputedProperty(obj, name, fn) {

  Object.defineProperty(obj, name, {
    configurable: true,
    get() {
      const getter = computed(fn, { context: this });
      Object.defineProperty(this, name, {
        configurable: false,
        get: () => getter.get(),
      });
      return getter.get();
    },
  });

}

export function extendHasMany(properties) {
  return (collection) => {
    forIn(properties, (fn, prop) => {
      addComputedProperty(collection, prop, fn);
    });
  };
}
