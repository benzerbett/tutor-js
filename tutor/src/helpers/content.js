import _ from 'underscore';


export default {
  // Convert each number to base 10 with it's position based on index.
  // If section is not present, 0 is set for it
  //   1 becomes 100, 1.1 becomes 101, 3.2 to 302, 3.2.1 -> 30201
  // Useful for sorting
  chapterSectionToNumber(chapter_section) {
    if (_.isString(chapter_section)) { chapter_section = chapter_section.split('.'); }
    if (chapter_section.length === 1) { chapter_section.push(0); } // add a section 0 if it has only a chapter
    let position = -1;
    return _.reduceRight(chapter_section, (memo, num) => memo + (num * Math.pow(100, (position += 1)))
      , 0);
  },
};
