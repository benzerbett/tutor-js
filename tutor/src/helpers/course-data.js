import Courses from '../models/courses-map';

import Router from '../helpers/router';
import TimeHelper from '../helpers/time';

export default {
  getCourseDataProps(courseId) {
    let dataProps;
    if (courseId == null) {
      ((((((({ courseId } = Router.currentParams())))))));
    }
    const course = Courses.get(courseId);
    return dataProps = {
      'data-title': course.name,
      'data-book-title': course.bookName || '',
      'data-appearance': course.appearance_code,
    };
  },

  getCourseBounds(courseId) {
    const course = Courses.get(courseId);

    const start = TimeHelper.getMomentPreserveDate(course.starts_at, TimeHelper.ISO_DATE_FORMAT);
    const end = TimeHelper.getMomentPreserveDate(course.ends_at, TimeHelper.ISO_DATE_FORMAT);

    return { start, end };
  },
};
