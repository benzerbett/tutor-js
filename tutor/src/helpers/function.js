export default {
  // coffeelint: disable=no_empty_functions
  empty() {},
};
// coffeelint: enable=no_empty_functions
