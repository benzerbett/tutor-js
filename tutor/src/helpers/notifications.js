import { NotificationActions } from 'shared';

export default {
  start(bootstrapData) {
    NotificationActions.startPolling();
    return (() => {
      const result = [];
      const object = bootstrapData.flash || {};
      for (let level in object) {
        const message = object[level];
        result.push(NotificationActions.display({ message, level }));
      }
      return result;
    })();
  },

  buildCallbackHandlers(comp) {
    // Require router here because requiring it in global scope
    // causes hot module reloading to fail in dev mode because this file is required outside the HMR scope
    const Router = require('./router');
    const { router } = comp.context;
    if (!router) {
      throw new Error('Component\'s context must have router present');
    }
    return {
      missing_student_id: {
        onAdd({ course }) {
          return router.history.push(
            Router.makePathname('changeStudentId', { courseId: course.id })
          );
        },
      },
    };
  },
};
