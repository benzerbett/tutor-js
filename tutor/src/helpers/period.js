import { filter, isEmpty, sortBy, toNumber, isNumber } from 'lodash';
import { addComputedProperty } from './computed-property';
import S from './string';

// Used to filter periods by helper methods
const isArchivedCheckFn = period => period.is_archived;
const isActiveCheckFn   = period => !period.is_archived;

const NumberLike = /[^0-9]+/;


const PeriodHelper = {
  getOrder(period) {
    return S.getNumberAndStringOrder(period.name);
  },

  sort(periods) {
    // expects either numbers, names with numbers or just names
    periods = sortBy(periods, (period) => { // eslint-disable-line consistent-return
      if (!isNumber(period.name)) {
        const name = period.name.match(/[^0-9]+/ig);
        if (name) {
          return name;
        }
      }
    });
    return sortBy(periods, (period) => { // eslint-disable-line consistent-return
      if (!isNumber(period.name)) {
        const number = period.name.match(/[0-9.-]+/g);
        if (number) {
          return parseFloat(number);
        }
      } else {
        return period.name;
      }
    });
  },

};


export default PeriodHelper;
