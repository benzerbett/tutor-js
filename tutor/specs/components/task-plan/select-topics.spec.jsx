import { React, SnapShot } from 'helpers';
import Factory, { FactoryBot } from '../../factories';
import SelectTopics from '../../../src/components/task-plan/select-topics';
import HOMEWORK from '../../../api/plans/2.json';

describe('Select Topics', function() {
  let props, course;

  beforeEach(() => {
    const course = Factory.course();
    course.referenceBook.onApiRequestComplete({
      data: [FactoryBot.create('Book')],
    });

    props = {
      course,
      planId: HOMEWORK.id,
      onSectionChange: jest.fn(),
      hide: jest.fn(),
      selected: [],
      header: 'Section Chooser Header',
      cancel: jest.fn(),
    };
  });

  it('matches snapshot', function() {
    expect.snapshot(<SelectTopics {...props} />).toMatchSnapshot();
  });

});
