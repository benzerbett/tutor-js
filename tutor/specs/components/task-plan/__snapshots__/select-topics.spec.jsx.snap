// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select Topics matches snapshot 1`] = `
<div
  className="undefined-plan-select-topics"
  data-tour-region-id="add-undefined-choose-sections"
>
  <div
    className="dialog default-dialog select-topics panel panel-default"
    id={undefined}
  >
    <div
      className="panel-heading"
    >
      Section Chooser Header
      <button
        className="openstax-close-x close"
        onClick={[Function]}
      />
    </div>
    <div
      className="panel-body"
    >
      <div
        className="select-chapters"
      >
        <div
          className="sections-chooser"
        >
          <div
            className="panel-group"
            role="tablist"
          >
            <div
              className="panel panel-default"
              id={null}
            >
              <div
                className="panel-heading"
              >
                <div
                  className="chapter-heading panel-title"
                  data-chapter-section={1}
                >
                  <a
                    aria-controls={undefined}
                    aria-expanded={true}
                    aria-selected={true}
                    className={null}
                    href={undefined}
                    onClick={[Function]}
                    role="tab"
                  >
                    <span
                      className="chapter-checkbox"
                    >
                      <span
                        className="tri-state-checkbox unchecked"
                        onClick={[Function]}
                        tabIndex={1}
                      >
                        <i
                          className="tutor-icon fa fa-square-o clickable"
                          onClick={[Function]}
                          role="button"
                          style={
                            Object {
                              "cursor": "pointer",
                            }
                          }
                          type="square-o"
                        />
                      </span>
                    </span>
                    <span
                      className="chapter-number"
                    >
                      Chapter 
                      <span
                        className="chapter-section"
                        data-chapter-section="1"
                      >
                        1
                      </span>
                       - 
                    </span>
                    <span
                      className="chapter-title"
                    >
                       
                      Kinematics
                       
                    </span>
                    <div
                      aria-label="Browse course reference book"
                      className="browse-the-book"
                      onClick={[Function]}
                      role="link"
                    >
                      Browse this Chapter
                    </div>
                  </a>
                </div>
              </div>
              <div
                aria-expanded={null}
                aria-hidden={false}
                className="panel-collapse collapse in"
                id={undefined}
                role="tabpanel"
              >
                <div
                  className="panel-body"
                >
                  <div
                    className="section"
                    data-page-id="234"
                    onClick={[Function]}
                  >
                    <span
                      className="section-checkbox"
                    >
                      <input
                        checked={false}
                        readOnly={true}
                        type="checkbox"
                      />
                    </span>
                    <span
                      className="chapter-section"
                      data-chapter-section="1.1"
                    >
                      1.1
                    </span>
                    <span
                      className="section-title"
                    >
                       
                      Kinematics in 1 dim
                    </span>
                  </div>
                  <div
                    className="section"
                    data-page-id="235"
                    onClick={[Function]}
                  >
                    <span
                      className="section-checkbox"
                    >
                      <input
                        checked={false}
                        readOnly={true}
                        type="checkbox"
                      />
                    </span>
                    <span
                      className="chapter-section"
                      data-chapter-section="1.2"
                    >
                      1.2
                    </span>
                    <span
                      className="section-title"
                    >
                       
                      Kinematics in 2 dim
                    </span>
                  </div>
                  <div
                    className="section"
                    data-page-id="236"
                    onClick={[Function]}
                  >
                    <span
                      className="section-checkbox"
                    >
                      <input
                        checked={false}
                        readOnly={true}
                        type="checkbox"
                      />
                    </span>
                    <span
                      className="chapter-section"
                      data-chapter-section="1.3"
                    >
                      1.3
                    </span>
                    <span
                      className="section-title"
                    >
                       
                      Kinematics in 3 dim
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            className="panel-group"
            role="tablist"
          >
            <div
              className="panel panel-default"
              id={null}
            >
              <div
                className="panel-heading"
              >
                <div
                  className="chapter-heading panel-title"
                  data-chapter-section={2}
                >
                  <a
                    aria-controls={undefined}
                    aria-expanded={false}
                    aria-selected={false}
                    className="collapsed"
                    href={undefined}
                    onClick={[Function]}
                    role="tab"
                  >
                    <span
                      className="chapter-checkbox"
                    >
                      <span
                        className="tri-state-checkbox unchecked"
                        onClick={[Function]}
                        tabIndex={1}
                      >
                        <i
                          className="tutor-icon fa fa-square-o clickable"
                          onClick={[Function]}
                          role="button"
                          style={
                            Object {
                              "cursor": "pointer",
                            }
                          }
                          type="square-o"
                        />
                      </span>
                    </span>
                    <span
                      className="chapter-number"
                    >
                      Chapter 
                      <span
                        className="chapter-section"
                        data-chapter-section="2"
                      >
                        2
                      </span>
                       - 
                    </span>
                    <span
                      className="chapter-title"
                    >
                       
                      Mechanics
                       
                    </span>
                    <div
                      aria-label="Browse course reference book"
                      className="browse-the-book"
                      onClick={[Function]}
                      role="link"
                    >
                      Browse this Chapter
                    </div>
                  </a>
                </div>
              </div>
              <div
                aria-expanded={null}
                aria-hidden={true}
                className="panel-collapse collapse"
                id={undefined}
                role="tabpanel"
              >
                <div
                  className="panel-body"
                >
                  <div
                    className="section"
                    data-page-id="237"
                    onClick={[Function]}
                  >
                    <span
                      className="section-checkbox"
                    >
                      <input
                        checked={false}
                        readOnly={true}
                        type="checkbox"
                      />
                    </span>
                    <span
                      className="chapter-section"
                      data-chapter-section="2.1"
                    >
                      2.1
                    </span>
                    <span
                      className="section-title"
                    >
                       
                      Mechanics in 1 dim
                    </span>
                  </div>
                  <div
                    className="section"
                    data-page-id="238"
                    onClick={[Function]}
                  >
                    <span
                      className="section-checkbox"
                    >
                      <input
                        checked={false}
                        readOnly={true}
                        type="checkbox"
                      />
                    </span>
                    <span
                      className="chapter-section"
                      data-chapter-section="2.2"
                    >
                      2.2
                    </span>
                    <span
                      className="section-title"
                    >
                       
                      Mechanics in 2 dim
                    </span>
                  </div>
                  <div
                    className="section"
                    data-page-id="239"
                    onClick={[Function]}
                  >
                    <span
                      className="section-checkbox"
                    >
                      <input
                        checked={false}
                        readOnly={true}
                        type="checkbox"
                      />
                    </span>
                    <span
                      className="chapter-section"
                      data-chapter-section="2.3"
                    >
                      2.3
                    </span>
                    <span
                      className="section-title"
                    >
                       
                      Mechanics in 3 dim
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            className="panel-group"
            role="tablist"
          >
            <div
              className="panel panel-default"
              id={null}
            >
              <div
                className="panel-heading"
              >
                <div
                  className="chapter-heading panel-title"
                  data-chapter-section={3}
                >
                  <a
                    aria-controls={undefined}
                    aria-expanded={false}
                    aria-selected={false}
                    className="collapsed"
                    href={undefined}
                    onClick={[Function]}
                    role="tab"
                  >
                    <span
                      className="chapter-checkbox"
                    >
                      <span
                        className="tri-state-checkbox unchecked"
                        onClick={[Function]}
                        tabIndex={1}
                      >
                        <i
                          className="tutor-icon fa fa-square-o clickable"
                          onClick={[Function]}
                          role="button"
                          style={
                            Object {
                              "cursor": "pointer",
                            }
                          }
                          type="square-o"
                        />
                      </span>
                    </span>
                    <span
                      className="chapter-number"
                    >
                      Chapter 
                      <span
                        className="chapter-section"
                        data-chapter-section="3"
                      >
                        3
                      </span>
                       - 
                    </span>
                    <span
                      className="chapter-title"
                    >
                       
                      Thermodynamics
                       
                    </span>
                    <div
                      aria-label="Browse course reference book"
                      className="browse-the-book"
                      onClick={[Function]}
                      role="link"
                    >
                      Browse this Chapter
                    </div>
                  </a>
                </div>
              </div>
              <div
                aria-expanded={null}
                aria-hidden={true}
                className="panel-collapse collapse"
                id={undefined}
                role="tabpanel"
              >
                <div
                  className="panel-body"
                >
                  <div
                    className="section"
                    data-page-id="240"
                    onClick={[Function]}
                  >
                    <span
                      className="section-checkbox"
                    >
                      <input
                        checked={false}
                        readOnly={true}
                        type="checkbox"
                      />
                    </span>
                    <span
                      className="chapter-section"
                      data-chapter-section="3.1"
                    >
                      3.1
                    </span>
                    <span
                      className="section-title"
                    >
                       
                      Thermodynamics in 1 dim
                    </span>
                  </div>
                  <div
                    className="section"
                    data-page-id="241"
                    onClick={[Function]}
                  >
                    <span
                      className="section-checkbox"
                    >
                      <input
                        checked={false}
                        readOnly={true}
                        type="checkbox"
                      />
                    </span>
                    <span
                      className="chapter-section"
                      data-chapter-section="3.2"
                    >
                      3.2
                    </span>
                    <span
                      className="section-title"
                    >
                       
                      Thermodynamics in 2 dim
                    </span>
                  </div>
                  <div
                    className="section"
                    data-page-id="242"
                    onClick={[Function]}
                  >
                    <span
                      className="section-checkbox"
                    >
                      <input
                        checked={false}
                        readOnly={true}
                        type="checkbox"
                      />
                    </span>
                    <span
                      className="chapter-section"
                      data-chapter-section="3.2"
                    >
                      3.2
                    </span>
                    <span
                      className="section-title"
                    >
                       
                      Thermodynamics in 3 dim
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="panel-footer"
    >
      <button
        className="btn btn-default"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Cancel
      </button>
    </div>
  </div>
</div>
`;
